vllm/v1/structured_output/backend_xgrammar.py

```diff
@@ -87,11 +88,22 @@ def __init__(self, vllm_config: VllmConfig):
             )
         self.compiler = xgr.GrammarCompiler(
             tokenizer_info,
-            max_threads=8,
+            max_threads=self._get_grammar_compiler_threads(),
             cache_enabled=True,
             cache_limit_bytes=vllm.envs.VLLM_XGRAMMAR_CACHE_MB * 1024 * 1024,
         )
 
+    def _get_grammar_compiler_threads(self) -> int:
+        """Get the number of threads to use for the grammar compiler. The number of physical
+        threads is used to determine the thread count, and it is capped at 32 because having too
+        many threads does not yield additional benefits.
```

Comment:
Do you think there's something special about 32? or just that it happened to be the number of physical cores on a single CPU in the test environment?

Reply:
The maximum number of effective threads is the number of elements in the grammar, typically in dozens. So 32 is a reasonable number for not very complex schemas. If the grammar is extremely complex, more effective threads would be allowed, but for general purposes I think we can fix it to 32.